### PR TITLE
[CELEBORN-1513] Support wildcard bind in dual stack environments

### PR DESCRIPTION
**Pull request by akpatnam25**
### What changes were proposed in this pull request?
Support wildcard bind for RPC and HTTP servers. When wildcard address is used, the service is able to listen to both ipv4 and ipv6 traffic in dual-stack environments. 

The specific scenario where this becomes relevant is as follows:

If some of the compute infrastructure is IPv4 only, some v6 only and others dual stack - the way we can have a single Celeborn infra to cater to all is by:
a) Set bind.preferip to false - so that advertised address is the host and not IP.

b) bind to wild card address

With both in place, the v4 only compute nodes will resolve the v4 address and connect to v4 ip/port.
Likewise, for v6 only.
Dual stack compute nodes will use prefer ipv6 Java flag to resolve to either v4 or v6.

This is how we are handling the combination of scenarios internally.

### Why are the changes needed?
see above. 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Tested on a server using netstat, and tried connecting to via `nc -4` and `nc -6` to ensure connection was there.

### CODE REVIEW

**Comment by akpatnam25** [comment]
this is a duplicate of https://github.com/apache/celeborn/pull/2633 -- I had forgotten to update it for a long time and it became stale with a lot of conflicts, hence I just made a new branch/PR.

**Comment by akpatnam25** [comment]
cc @FMX @SteNicholas @pan3793 @RexXiong @turboFei please review, thank you :)

File: common/src/main/scala/org/apache/celeborn/common/util/Utils.scala [hunk 1]
```diff
@@ -415,8 +415,21 @@ object Utils extends Logging {
...
   }
 
-  def localHostName(conf: CelebornConf): String = customHostname.getOrElse {
-    if (conf.bindPreferIP) {
+  def localHostName(conf: CelebornConf): String = {
+    getHostName(conf.bindPreferIP)
+  }
+
+  def localHostNameForAdvertiseAddress(conf: CelebornConf): String = {
+    if (conf.masterHost != "<localhost>") {
```
Comment (RexXiong) [inline common/src/main/scala/org/apache/celeborn/common/util/Utils.scala:423]: We should separate the logic for obtaining the AdvertiseAddress for the master and worker nodes.

Comment (akpatnam25) [reply common/src/main/scala/org/apache/celeborn/common/util/Utils.scala:432]: fixed

**Comment by akpatnam25** [comment]
thanks @RexXiong , I have addressed your comments. Please take another look , thanks!! :)

File: common/src/main/scala/org/apache/celeborn/common/util/Utils.scala [hunk 2]
```diff
@@ -415,8 +415,33 @@ object Utils extends Logging {
...
+        RpcNameConstants.MASTER_INTERNAL_SYS)) {
+      getAdvertiseAddressForMaster(conf)
+    } else {
+      getAdvertiseAddressForWorker(conf)
+    }
+  }
+
+  private def getAdvertiseAddressForMaster(conf: CelebornConf) = {
+    if (conf.masterHost != "<localhost>") {
+      conf.masterHost
```
Comment (RexXiong) [inline common/src/main/scala/org/apache/celeborn/common/util/Utils.scala:433]: It seems that conf.masterHost is never equal to "<localhost>". Why not directly call getHostName(conf.advertisePreferIP) instead?

Comment (akpatnam25) [reply common/src/main/scala/org/apache/celeborn/common/util/Utils.scala:433]: this is to cover this unit test case: https://github.com/apache/celeborn/blob/main/worker/src/test/scala/org/apache/celeborn/service/deploy/MiniClusterFeature.scala#L61

Comment (akpatnam25) [reply common/src/main/scala/org/apache/celeborn/common/util/Utils.scala:433]: the check should maybe be: 
```
    if (conf.masterHost == "localhost") {
      conf.masterHost
    } else {
      getHostName(conf.advertisePreferIP)
    }
```
@RexXiong wdyt?

Comment (RexXiong) [reply common/src/main/scala/org/apache/celeborn/common/util/Utils.scala:433]: Perhaps you mean that if celeborn.master.host is set, we should prioritize using that. If that’s the case, we can define a new method for this.

```
      def advertiseMasterHost: String = get(MASTER_HOST).replace(<localhost>", Utils.getHostName(this.advertisePreferIP))
```

Comment (akpatnam25) [reply common/src/main/scala/org/apache/celeborn/common/util/Utils.scala]: sure i am fine with this

Comment (akpatnam25) [reply common/src/main/scala/org/apache/celeborn/common/util/Utils.scala]: updated

**Review by RexXiong — APPROVED** [review]

**Review by SteNicholas — APPROVED** [review]
LGTM.

**Comment by akpatnam25** [comment]
@RexXiong @SteNicholas  are we able to merge this? :)

**Review by FMX — APPROVED** [review]
LGTM. Merged into main(0.6.0).

**Comment by akpatnam25** [comment]
thanks everyone for the reviews!!

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 3]
```diff
@@ -1723,6 +1728,23 @@ object CelebornConf extends Logging {
       .intConf
       .createOptional
 
+  val NETWORK_WILDCARD_ADDRESS_BIND: ConfigEntry[Boolean] =
+    buildConf("celeborn.network.bind.wildcardAddress")
+      .categories("network")
+      .version("0.6.0")
+      .doc("When `true`, the bind address will be set to a wildcard address, while the advertise address will " +
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1735]: Is "wildcard address" an idiomatic networking term used to represent an arbitrary address(`0.0.0.0` in IPv4 case and `::0` in IPv6)? I also saw [a similar question](https://github.com/apache/celeborn/pull/2633#discussion_r1690084955) in the previous PR.

I'm not sure what the official or most popular names of `0.0.0.0` and `::0` are, according to https://en.wikipedia.org/wiki/0.0.0.0, the most common one might be `INADDR_ANY`.

Comment (pan3793) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1735]: I believe "wildcard address" with the current docs would confuse some users (including me), based on the results returned by the search engines (Google and Bing) "anyAddress" might be a better option. Anyway, please at least enrich the doc of this configuration.

Comment (akpatnam25) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1735]: according to InetAddress docs it is defined as `wildcard address`: https://docs.oracle.com/javase/7/docs/api/java/net/InetSocketAddress.html#InetSocketAddress(int) 
I can make the doc more clear, thanks for bringing this up @pan3793

Comment (pan3793) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1735]: okay, let's keep the name and update the doc then.

**Pull request closed without merging**